File: sdk/javascript/ServiceBroker.ts
import {
  ChaincodeEventsOptions,
  CloseableAsyncIterable,
  Network,
} from '@hyperledger/fabric-gateway';
import { TextDecoder } from 'util';

import Contract, { ContractInterface } from './Contract';
import ServiceRequest from './ServiceRequest';
import ServiceRequestResponse from './ServiceRequestResponse';
import ServiceResponse from './ServiceResponse';

/**
 * Interface of the event emitted by the service broker contract notifying a service request/response update
 */
export interface ServiceRequestEvent {
  /**
   * Name of the action performed on the service request
   */
  action: string;

  /**
   * Organization ID of the requested service
   */
  organizationId: string;

  /**
   * Device ID of the requested service
   */
  deviceId: string;

  /**
   * Name of the requested service
   */
  serviceName: string;

  /**
   * ID of the request
   */
  requestId: string;

  /**
   * Custom event payload
   */
  payload: any;
}

/**
 * Interface of core utilities for managing service requests on ledger
 */
export interface ServiceBrokerInterface {
  /**
   * Make a request to an IoT service
   *
   * @param request IoT service request to be sent
   */
  request(request: ServiceRequest): Promise<void>;

  /**
   * Respond to an IoT service request
   *
   * @param response IoT service response to be sent
   */
  respond(response: ServiceResponse): Promise<void>;

  /**
   * Get an IoT service request and its response (if any) by the request ID
   *
   * @param requestId service request ID
   * @returns the service request and its response (if any)
   */
  get(requestId: string): Promise<ServiceRequestResponse>;

  /**
   * Get a list of IoT service requests and their responses (if any) by their service organization ID, service device ID, and service name
   *
   * @param organizationId organization ID of the requested service
   * @param deviceId device's ID of the requested service
   * @param serviceName name of the requested service
   * @returns all services of the device in the organization
   */
  getAll(
    organizationId: string,
    deviceId: string,
    serviceName: string,
  ): Promise<ServiceRequestResponse[]>;

  /**
   * Remove a service request and its response (if any) from the ledger
   *
   * @param requestId ID of the service request and response to be removed
   */
  remove(requestId: string): Promise<void>;

  /**
   * Registers for service request events
   *
   * @param options chaincode event options
   * @returns an async iterable of the events
   */
  registerEvent(
    options: ChaincodeEventsOptions,
  ): Promise<CloseableAsyncIterable<ServiceRequestEvent>>;
}

/**
 * Core utilities for managing IoT service requests and responses on the ledger
 */
export default class ServiceBroker implements ServiceBrokerInterface {
  private utf8Decoder: TextDecoder;

  /**
   * @param contract smart contract
   */
  constructor(private contract: ContractInterface) {
    this.utf8Decoder = new TextDecoder();
  }

  async request(request: ServiceRequest): Promise<void> {
    const serialized = request.serialize();
    await this.contract.submitTransaction('Request', serialized);
  }

  async respond(response: ServiceResponse): Promise<void> {
    const serialized = response.serialize();
    await this.contract.submitTransaction('Respond', serialized);
  }

  async get(requestId: string): Promise<ServiceRequestResponse> {
    const data = await this.contract.submitTransaction('Get', requestId);
    const serialized = this.utf8Decoder.decode(data);
    return ServiceRequestResponse.deserialize(serialized);
  }

  async getAll(
    organizationId: string,
    deviceId: string,
    serviceName: string,
  ): Promise<ServiceRequestResponse[]> {
    const data = await this.contract.submitTransaction(
      'GetAll',
      organizationId,
      deviceId,
      serviceName,
    );
    const serialized = this.utf8Decoder.decode(data);
    return (JSON.parse(serialized) as any[]).map(ServiceRequestResponse.fromObject);
  }

  async remove(requestId: string): Promise<void> {
    await this.contract.submitTransaction('Remove', requestId);
  }

  async registerEvent(
    options?: ChaincodeEventsOptions,
  ): Promise<CloseableAsyncIterable<ServiceRequestEvent>> {
    const events = await this.contract.registerEvent(options);
    const pattern = /^request:\/\/(.+?)\/(.+?)\/(.+?)\/(.+?)\/(.+?)$/;
    const decoder = this.utf8Decoder;

    return {
      async *[Symbol.asyncIterator]() {
        for await (const event of events) {
          // can reuse pattern here since it has no global('g') flag
          const matches = pattern.exec(event.eventName);
          if (matches === null || matches.length != 6) {
            continue;
          }

          const serviceRequestEvent: ServiceRequestEvent = {
            organizationId: matches[1],
            deviceId: matches[2],
            serviceName: matches[3],
            requestId: matches[4],
            action: matches[5],
            payload: null,
          };

          if (serviceRequestEvent.action === 'request') {
            try {
              const request = ServiceRequest.deserialize(decoder.decode(event.payload));
              serviceRequestEvent.payload = request;
            } catch {
              console.error(
                `bad service request event payload ${event.payload}, action is ${serviceRequestEvent.action}`,
              );
              continue;
            }
          } else if (serviceRequestEvent.action === 'respond') {
            try {
              const respond = ServiceResponse.deserialize(decoder.decode(event.payload));
              serviceRequestEvent.payload = respond;
            } catch {
              console.error(
                `bad service response event payload ${event.payload}, action is ${serviceRequestEvent.action}`,
              );
              continue;
            }
          } else if (serviceRequestEvent.action === 'remove') {
            serviceRequestEvent.payload = decoder.decode(event.payload);
          } else {
            serviceRequestEvent.payload = event.payload;
          }

          yield serviceRequestEvent;
        }
      },
      close() {
        events.close();
      },
    };
  }
}

export function createServiceBroker(network: Network, chaincodeId: string) {
  return new ServiceBroker(new Contract(network, chaincodeId, 'service_broker'));
}

File: sdk/go/service_broker.go
package sdk

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"regexp"

	"github.com/hyperledger/fabric-gateway/pkg/client"
	"github.com/nexus-lab/iot-service-blockchain/common"
)

// ServiceRequestEvent an event emitted by the service broker contract notifying a service request/response update
type ServiceRequestEvent struct {
	// Action name of the action performed on the service request
	Action string

	// OrganizationId organization ID of the requested service
	OrganizationId string

	// DeviceId device ID of the requested service
	DeviceId string

	// ServiceName name of the requested service
	ServiceName string

	// RequestId ID of the request
	RequestId string

	// Payload custom event payload
	Payload interface{}
}

// ServiceBrokerInterface core utilities for managing service requests on ledger
type ServiceBrokerInterface interface {
	// Request make a request to an IoT service
	Request(request *common.ServiceRequest) error

	// Respond respond to an IoT service request
	Respond(response *common.ServiceResponse) error

	// Get return an IoT service request and its response (if any) by the request ID
	Get(requestId string) (*common.ServiceRequestResponse, error)

	// GetAll return a list of IoT service requests and their responses (if any) by their service organization ID, service device ID, and service name
	GetAll(organizationId string, deviceId string, serviceName string) ([]*common.ServiceRequestResponse, error)

	// Remove remove a service request and its response (if any) from the ledger
	Remove(requestId string) error

	// RegisterEvent registers for service request events
	RegisterEvent(options ...client.ChaincodeEventsOption) (<-chan *ServiceRequestEvent, context.CancelFunc, error)
}

// ServiceBroker core utilities for managing IoT service requests and responses on the ledger
type ServiceBroker struct {
	contract ContractInterface
}

// Request make a request to an IoT service
func (r *ServiceBroker) Request(request *common.ServiceRequest) error {
	if request == nil {
		return fmt.Errorf("cannot send an empty request")
	}

	data, err := request.Serialize()
	if err != nil {
		return err
	}

	_, err = r.contract.SubmitTransaction("Request", string(data))
	return err
}

// Respond respond to an IoT service request
func (r *ServiceBroker) Respond(response *common.ServiceResponse) error {
	if response == nil {
		return fmt.Errorf("cannot send an empty response")
	}

	data, err := response.Serialize()
	if err != nil {
		return err
	}

	_, err = r.contract.SubmitTransaction("Respond", string(data))
	return err
}

// Get return an IoT service request and its response by the request ID
func (r *ServiceBroker) Get(requestId string) (*common.ServiceRequestResponse, error) {
	data, err := r.contract.SubmitTransaction("Get", requestId)
	if err != nil {
		return nil, err
	}

	return common.DeserializeServiceRequestResponse(data)
}

// GetAll return a list of IoT service requests and their responses by their organization ID, device ID, and service name
func (r *ServiceBroker) GetAll(organizationId string, deviceId string, serviceName string) ([]*common.ServiceRequestResponse, error) {
	data, err := r.contract.SubmitTransaction("GetAll", organizationId, deviceId, serviceName)
	if err != nil {
		return nil, err
	}

	results := make([]*common.ServiceRequestResponse, 0)
	if err = json.Unmarshal(data, &results); err != nil {
		return nil, err
	}

	return results, nil
}

// Remove remove a (request, response) pair from the ledger
func (r *ServiceBroker) Remove(requestId string) error {
	_, err := r.contract.SubmitTransaction("Remove", requestId)
	return err
}

// RegisterEvent registers for service request events
func (r *ServiceBroker) RegisterEvent(options ...client.ChaincodeEventsOption) (<-chan *ServiceRequestEvent, context.CancelFunc, error) {
	dest := make(chan *ServiceRequestEvent)
	source, cancel, err := r.contract.RegisterEvent(options...)
	pattern := regexp.MustCompile(`^request:\/\/(.+?)\/(.+?)\/(.+?)\/(.+?)\/(.+?)$`)

	go func() {
		defer close(dest)

		for event := range source {
			matches := pattern.FindStringSubmatch(event.EventName)
			if len(matches) != 6 {
				continue
			}

			serviceRequestEvent := &ServiceRequestEvent{
				OrganizationId: matches[1],
				DeviceId:       matches[2],
				ServiceName:    matches[3],
				RequestId:      matches[4],
				Action:         matches[5],
			}

			if serviceRequestEvent.Action == "request" {
				request, err := common.DeserializeServiceRequest(event.Payload)
				if err != nil {
					log.Printf("bad service request event payload %#v, action is %s\n", event.Payload, serviceRequestEvent.Action)
					continue
				}
				serviceRequestEvent.Payload = request
			} else if serviceRequestEvent.Action == "respond" {
				response, err := common.DeserializeServiceResponse(event.Payload)
				if err != nil {
					log.Printf("bad service response event payload %#v, action is %s\n", event.Payload, serviceRequestEvent.Action)
					continue
				}
				serviceRequestEvent.Payload = response
			} else if serviceRequestEvent.Action == "remove" {
				serviceRequestEvent.Payload = string(event.Payload)
			} else {
				serviceRequestEvent.Payload = event.Payload
			}

			dest <- serviceRequestEvent
		}
	}()

	return dest, cancel, err
}

// CreateServiceBroker the default factory for creating service brokers
func CreateServiceBroker(network *client.Network, chaincodeId string) ServiceBrokerInterface {
	return &ServiceBroker{
		contract: &Contract{
			network:      network,
			chaincodeId:  chaincodeId,
			contractName: "service_broker",
		},
	}
}
